clippy_lints/src/string_patterns.rs

```diff
@@ -0,0 +1,239 @@
+            hint,
+            applicability,
+        );
+    }
+}
+
+fn get_char_value(expr: &Expr<'_>) -> Option<String> {
+    match expr.kind {
+        ExprKind::Lit(lit) => match lit.node {
+            LitKind::Char(c) => Some(format!("'{}'", c.escape_default())),
+            _ => None,
+        },
+        ExprKind::Path(QPath::Resolved(_, path)) => {
+            if path.segments.len() == 1 {
+                let segment = &path.segments[0];
+                if segment.args.is_none() {
+                    Some(segment.ident.name.to_string())
+                } else {
+                    None
+                }
+            } else {
+                None
+            }
+        },
+        _ => None,
+    }
+}
+
+fn check_manual_pattern_char_comparison(cx: &LateContext<'_>, method_arg: &Expr<'_>) {
+    let applicability = Applicability::MachineApplicable;
+    if let ExprKind::Closure(closure) = method_arg.kind
+        && let body = cx.tcx.hir().body(closure.body)
+        && let PatKind::Binding(_, binding, ..) = body.params[0].pat.kind
+    {
+        let mut set_chars: Vec<String> = Vec::new();
+
+        // We want to retrieve all the comparisons done.
+        // They are ordered in a nested way and so we need to traverse the AST to collect them all.
+        if for_each_expr(body.value, |sub_expr| -> ControlFlow<(), Descend> {
+            match sub_expr.kind {
+                ExprKind::Binary(op, left, right) if op.node == BinOpKind::Eq => {
+                    if path_to_local_id(left, binding)
+                        && cx.typeck_results().expr_ty_adjusted(right).kind() == &ty::Char
+                        && let Some(c) = get_char_value(right)
+                    {
+                        set_chars.push(c);
+                        ControlFlow::Continue(Descend::No)
+                    } else if path_to_local_id(right, binding)
+                        && cx.typeck_results().expr_ty_adjusted(left).kind() == &ty::Char
+                        && let Some(c) = get_char_value(left)
+                    {
+                        set_chars.push(c);
+                        ControlFlow::Continue(Descend::No)
+                    } else {
+                        ControlFlow::Break(())
+                    }
+                },
+                ExprKind::Binary(op, _, _) if op.node == BinOpKind::Or => ControlFlow::Continue(Descend::Yes),
+                ExprKind::Match(match_value, [arm, _], _) => {
+                    if matching_root_macro_call(cx, sub_expr.span, sym::matches_macro).is_none() {
+                        return ControlFlow::Break(());
+                    }
+                    if arm.guard.is_some() {
+                        return ControlFlow::Break(());
+                    }
+                    if !path_to_local_id(match_value, binding) {
+                        return ControlFlow::Break(());
+                    }
+                    if arm.pat.walk_short(|pat| match pat.kind {
```

Comment:
These `if`s can be merged, or if it makes the indentation weird at least the first 3 can

Reply:
Done on the three first. The fourth one is more complex and I find that merge them isn't more readable.